### PR TITLE
Make OSA fields extendable, simplify naming

### PR DESCRIPTION
**Pull request by tomalec**
_This is an add-on to https://github.com/woocommerce/woocommerce/pull/39701_

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

- [Use `referrer` & `source_type` field names consistently](https://github.com/woocommerce/woocommerce/commit/1e293d9c4668b8f7f2081b65d44d753cc1440a1b) 
   Remove the need to translate it back and forth.
- [Make fields actually extendable](https://github.com/woocommerce/woocommerce/commit/8cfa6a2e6c452aa4a307a124977f7824365a5db5) 
   using `wc_order_attribution_tracking_fields`.
   Propagate the field configuration to the client side as well.
   
   Disambiguate `fields` in variables and functions.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Smoke test https://github.com/woocommerce/woocommerce/pull/39701 for regression
2. Add PHP snippet
```php

add_filter(
    'wc_order_attribution_tracking_fields',
    function( $fields ) {
        $fields['new_field'] = 'first_add.fd';
        return $fields;
    }
);
```
4. Visit classic checkout, 
   1. check that `<input type="hidden" name="wc_order_attribution_new_field" value="" />` is in the source code.
   2. Check its value is correctly set `document.querySelector('input[name="wc_order_attribution_new_field"]').value === sbjs.get.first_add.fd`
5. Visit Block checkout, and open dev tools
   1. click "Place order"
   2. Check `/checkout` ajax request payload contains the new field
      image
6. Check that order metadata contains the new property.
7. Review the code and check if all `field`-related names are understandable

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by message-dimke** [comment]
Hey, @tomalec !

I am curious what should the value be, `first_add.fd` or `sbjs.get.first_add.fd`, or a `date time`?
The filter sets  `first_add.fd` but in your test instructions for blocks the image shows `date time` and the actual tests show `date time`.

I am getting `date time` for both classical checkout and the block one.

Block:

Fullscreen_09_12_2023__18_19

Classic:

Fullscreen_09_12_2023__18_22

**Comment by tomalec** [comment]
> I am curious what should the value be, `first_add.fd` or `sbjs.get.first_add.fd`, or a `date time`?

The value of the `wc_order_attribution_tracking_fields` item should be the property accessor within sbjs.get object: `first_add.fd`

then when processed in browser's runtime it should be resolved to the value of "sbjs.get.the.accessor.you.have.given". Which in the example should be the value of `sbjs.get.first_add.fd`. Which per [sourcebuster docs ](https://sbjs.rocks/#/usage) is
>  Date and time of a visit. Format: `yyyy-mm-dd hh:mm:ss`...

So, the results you see are correct.

**Review by message-dimke — APPROVED** [review]
LGTM! Thank you, @tomalec !

File: plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php
Comment (tomalec) [inline plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php:427]: @layoutd If we want to keep the historic tracked data consistent, we can keep the property name as before, but still use the cleaned up field name:
```suggestion
			'type'                 => $source_data['source_type'] ?? '',
```

Comment (layoutd) [reply plugins/woocommerce/src/Internal/Orders/OrderAttributionController.php:427]: I think your solution is more elegant, having different terms for the same item will certainly lead to confusion in the future. Discussing it with @budzanowski, this change can be targeted for 8.6, and we can update all affected orders when the version is upgraded to avoid having to consider multiple keys.

File: plugins/woocommerce/templates/order/attribution-data-fields.php
Comment (layoutd) [inline plugins/woocommerce/templates/order/attribution-data-fields.php]: Changed since the PR was added.
```suggestion
 * @version 8.6.0
```

**Review by layoutd — APPROVED** [review]
These looks like they're working well! I tested with block and shortcode checkouts, and with several types of source types. The template version needs to be updated to match the current trunk version.

I was going to suggest we add metadata migration in another PR to allow sites that already have OA data work with these changes, but I realized that the OA feature already maps the type and url to `source_type` and `referrer` meta keys:
https://github.com/woocommerce/woocommerce/blob/2749cebf7f06b5dbfd45550a8dc4f65038ff9bb8/plugins/woocommerce/src/Internal/Traits/OrderAttributionMeta.php#L154-L170

So that's one fewer thing to do!

**Comment by tomalec** [comment]
I'm marking it as [high-impact](https://github.com/woocommerce/woocommerce/blob/trunk/docs/contributing/deciding-pr-high-impact.md#you-should-mark-a-pull-request-as-high-impact-if) as:

- It's a big PR (i.e. adds several changes in many files).
- Modifies critical functionality: Shopper | Checkout | Existing customer can place order
- Changes to WooCommerce hooks/actions/filters.

File: plugins/woocommerce/templates/order/attribution-data-fields.php
Comment (layoutd) [reply plugins/woocommerce/templates/order/attribution-data-fields.php]: I think this change should fix the tests (this was the only one still failing when I reviewed).

**Comment by alopezari** [comment]
Hi there! The E2E tests are failing here, I triggered a re-run just in case the failure was introduced by these changes. Will keep an eye on this.

**Update: They passed 👍**

**Comment by tomalec** [comment]
This PR is mostly with developer-facing tweaks. Also, may require some follow-up action to double-check if tracking- and meta- data migrate well.

As https://github.com/woocommerce/woocommerce/pull/43012 got full approval and analysis, while  https://github.com/woocommerce/woocommerce/pull/41856 (its version based on this PR) didn't, I merged #43012, to deliver the customer-facing fix ASAP.

The above means, `trunk` received changes that conflicts with this branch, and requires some adjustments. I tried to cover them in the merge commit, non-trivial changes were made to https://github.com/woocommerce/woocommerce/pull/41690/commits/e887ea32d7138be27548dfbe203f8c985532f436#diff-21bb08f041552efddbc9e2515315dddcedadedef055476d4fe4dd35962d5c25f

Maybe another pair of eyes could check if those changes didn't break anything?

**Pull request merged**
